### ISSUE TITLE
Users registering for an event can cause over-capacity situations

### ISSUE THREAD

**Issue by cabal95** [SparkDevNetwork/Rock#2510]
### Prerequisites
* [X] Put an X between the brackets on this line if you have done all of the following:
    * Can you reproduce the problem on a fresh install or the [demo site](http://rock.rocksolidchurchdemo.com/)?
    * Did you include your Rock version number and [client culture](https://github.com/SparkDevNetwork/Rock/wiki/Environment-and-Diagnostics-Information) setting?
    * Did you [perform a cursory search](https://github.com/issues?q=is%3Aissue+user%3ASparkDevNetwork+-repo%3ASparkDevNetwork%2FSlack) to see if your bug or enhancement is already reported?

### Description

When registering for an event, the capacity test and "past end-date" test are performed when the page is first loaded. This means if there is one spot left and two users begin the process of registering, they can both end up with a valid registration and cause the event to be over-capacity. This in itself is somewhat expected unless we enforce a hard limit and allow people to go through the entire process up to processing the CC payment and then check again and say "sorry, thanks for wasting your time but we are full now".

However, another related issue is that people can leave their browser window open for a long time (days perhaps?) in the middle of a registration and still complete it.

One suggestion is to introduce a "timeout counter" on the page that tells people how long they have left to complete the registration. This would at least ensure that the number of overages is minimized as people couldn't sit for hours on the page before completing the registration.

### Steps to Reproduce

1. Set event capacity to 1
2. Start two registration processes in two different browsers
3. Finish one.
4. Wait an hour or so, finish the second.
5. Notice that registrants is now 2 with capacity set to 1.

1. Set end date to noon.
2. Begin registration at 11am.
3. Complete registration at 4pm.

### Versions

* **Rock Version:** 1.6.8
* **Client Culture Setting:** en-US

**Comment by garrettjohnson** [comment]
Related to #2059

**Comment by edmistj** [comment]
We've talked through this in the past. The idea of a count down came up, but it would be complex to temporary lock out the number of requested seats. In the decade of doing large event registrations at CCV (and related input from others) it doesn't seem like a likely issue. While we'd like to add it some day there much lower fruit on the tree right now.

Waitlists do help with this a bit.

**Comment by cabal95** [comment]
Unfortunately for us we have been running into this situation fairly often - as best we can tell from people walking away from their browser and coming back hours later (which also seems to bypass the "registration end date"). But understand the complexity. Maybe we just have really distracted people here in LA. :)

What would be your thoughts to an **optional** event registration template setting of "hard limit" that did a final check just before submitting payment to gateway to see if this registration would put them over the limit? There would still be a narrow window of opportunity for over-capacity if two people hit submit at the same time, but it would be tight fit.  Our biggest concern is things like student camp where we have a specific number of spots available from the camp. If we go over we have to call somebody and tell them they can't go now, if we go under (for safety) we have to fork over money to pay for the spots anyway.

**Comment by edmistj** [comment]
You might check the code in v7. It's been over a year since we wrote the waitlist feature, but I think I remember a final check like that that would warn them they we're going on a waitlist.

**Comment by cabal95** [comment]
Okay thanks I'll take a look at that. My understanding was it did the waitlist "calculation" at the start of the registration process instead of at the end. I'll see if I can test our scenario on pre-alpha to see if that is the case or not. Thanks.

**Comment by edmistj** [comment]
I maybe dreaming but I thought it did a check at the end too. If not that might be the answer there. "Sorry, unfortunately there are only wait list spots available....'

**Comment by cabal95** [comment]
@edmistj Well, I tried testing on prealpha.rocksolidchurchdemo.com but ran into a few issues that prevented me from knowing anything further. I don't know if prealpha is an older release or if these are still bugs:

1. On the "how many" screen, the "Registration Full, will put on waitlist" message stayed up no matter what. I set the registration capacity for 2 people and it came up even when only registering one. I set the registration capacity for 10 people and the message still came up. Tried incementing and then decrementing the number of people to register and the message never went away.

2. No matter how many times I registered, it never put people on the waiting list or said anything further about the waiting list. For example, I set the capacity to 2 registrants. Then in browser A I started registering 2 people. Then in browser B started and completed registering 1 person. Back in browser A I finished the process and then had 3 registrants. **Then** in browser A I started a new registration (on the already over-capacity instance) and it let me register yet another person. So now I have 4 registrants across 3 registrations for the same event instance that only allows 2 people max. None of them went to the waiting list.

**Comment by cabal95** [comment]
Wait, scratch 85% of that. I was setting the max # of registrants that can be registered at one time not the instance capacity.  The first item is still a partial issue, it seems to display that message _always_ if the instance capacity is set to 0 (no limit). Testing again.

**Comment by edmistj** [comment]
OK, if you still find an issue please post it. Otherwise I'll assume all is well.

**Comment by cabal95** [comment]
Okay, ignore number 2 above.  The "how many" screen wait list message works correctly IF your instance capacity is > 0, otherwise it always shows up (bug). I can file a bug on that if you want, though as I said not sure how old the prealpha site is and if its already been fixed?

Same scenario as above, set _instance_ capacity to 2 this time.

* In browser A selected 2 on the how many screen and clicked next (so waiting to enter first registrant info).
* In browser B began and finished registration for 1 person. (instance now has 1 registrant)
* Back in browser A, entered information for first person (no warning about wait list).
* Continued and entered information for second person (still no warning about wait list).
* Summary screen shows and says 2 people will be registered (still no warning about wait list).
* Clicked Finish to complete registration, says 2 people were registered (instance now has 3 registrants).
* Checked in back-end, 3 people are registered, 0 on wait list.

So, maybe that is the discussion point for this whole thing. At the very least, re-check on the summary screen about wait list needs. Ideally, do a check when they click Finish (or Next to submit payment info) and display a message if the number of people that need to go on the wait list is different than the initial check. The message would say something like "The number of available spots has changed, the following people will be registered: A, B, C,; the following will be placed on the wait list: D".

**Comment by edmistj** [comment]
Can you summarize the bug described as #1 above. I've tested this and it works as I would expect. Setting the capacity to 0 with a wait list enabled **should** always show the message. This enables a wait list only event. But maybe I'm not understanding your report.

As to the other check, that could be added in the future. The problem would be the person would likely want to select which individuals to add to the event and which to place on the wait list ('No... I didn't want D on the wait list I wanted A') and the whole time this message is displayed (and perhaps a selection is being made) that final slot _could_ be taken by a different registration. 

Without locking seats with counters this will never be perfect, but that would add a lot of complexity both to the system and the registration process (personally I dislike the counters, I feel rushed and stressed watching them count down, this would especially be the case if I needed to fill out a large registration ala camp).

**Comment by cabal95** [comment]
> Setting the capacity to 0 with a wait list enabled should always show the message. This enables a wait list only event. But maybe I'm not understanding your report.

That makes sense. In that case I would say the bug is that it _didn't_ put me on the wait list, it only warned at the beginning. Steps to reproduce _on prealpha site_:

1. Create Registration Template with Wait List enabled.
2. Create Registration Instance with Capacity set to 0.
3. Register a person
4. Person sees initial "wait list" message on the "how many" screen, but never again during the registration.
5. Person is registered instead of put on the wait list.

My local v7/alpha dev environment is somewhat right now so I can't verify on latest git code at the moment, but the registration instance I built for testing on pre-alpha is http://prealpha.rocksolidchurchdemo.com/RegistrationInstance/3

**Comment by edmistj** [comment]
Ok that clarifies it. That probably deserves it's own Issue. Can you submit that? I did confirm it.
